ksql-parser/src/main/java/io/confluent/ksql/parser/rewrite/AstSanitizer.java

```diff
@@ -0,0 +1,283 @@
+      if (expression instanceof ColumnReferenceExp) {
+        final ColumnRef name = ((ColumnReferenceExp) expression).getReference();
+        if (dataSourceExtractor.isJoin()
+            && dataSourceExtractor.getCommonFieldNames().contains(name.name())) {
+          alias = name.source()
+              .map(q -> q.name() + "_" + name.name().name())
+              .map(ColumnName::of)
+              .orElseGet(name::name);
+        } else {
+          alias = name.name();
+        }
+      } else if (expression instanceof DereferenceExpression) {
+        final DereferenceExpression dereferenceExp = (DereferenceExpression) expression;
+        final String dereferenceExpressionString = dereferenceExp.toString();
+        alias = ColumnName.of(replaceDotFieldRef(
+            dereferenceExpressionString.substring(
+                dereferenceExpressionString.indexOf(KsqlConstants.DOT) + 1)));
+      } else {
+        alias = ColumnName.generatedColumnAlias(selectItemIndex);
+      }
+      selectItemIndex++;
+      return Optional.of(
+          new SingleColumn(singleColumn.getLocation(), expression, Optional.of(alias))
+      );
+    }
+
+    private static String replaceDotFieldRef(final String input) {
+      return input
+          .replace(KsqlConstants.DOT, "_")
+          .replace(KsqlConstants.STRUCT_FIELD_REF, "__");
+    }
+
+    private DataSource<?> getSource(
+        final SourceName name,
+        final Optional<NodeLocation> location
+    ) {
+      final DataSource<?> source = metaStore.getSource(name);
+      if (source == null) {
+        throw new InvalidColumnReferenceException(location, name.name() + " does not exist.");
+      }
+
+      return source;
+    }
+  }
+
+  private static final class ExpressionRewriterPlugin extends
+      VisitParentExpressionVisitor<Optional<Expression>, Context<Void>> {
+    final MetaStore metaStore;
+    final DataSourceExtractor dataSourceExtractor;
+
+    ExpressionRewriterPlugin(
+        final MetaStore metaStore,
+        final DataSourceExtractor dataSourceExtractor) {
+      super(Optional.empty());
+      this.metaStore = Objects.requireNonNull(metaStore, "metaStore");
+      this.dataSourceExtractor
+          = Objects.requireNonNull(dataSourceExtractor, "dataSourceExtractor");
+    }
+
+    @Override
+    public Optional<Expression> visitColumnReference(
+        final ColumnReferenceExp expression,
+        final Context<Void> ctx) {
+      final ColumnRef columnRef = expression.getReference();
+      if (columnRef.source().isPresent()) {
+        throwOnUnknownNameOrAlias(dataSourceExtractor, columnRef.source().get());
+        return Optional.empty();
+      }
+
+      final ColumnName columnName = columnRef.name();
+      if (dataSourceExtractor.isJoin()) {
+        if (dataSourceExtractor.getCommonFieldNames().contains(columnName)) {
+          throw new KsqlException("Column '" + columnName.name() + "' is ambiguous.");
```

Comment:
Such a check should be moved into the engine's execution path and out of the parser.